{
  "name": "example_script",
  "type": "CFG",
  "blocks": [
    {
      "id": 1,
      "start_line": 1,
      "end_line": 7,
      "label": "\"\"\"\nProblem source: https://www.hackerrank.com/challenges/the-power-sum/problem\nFind the number of ways that a given integer X, can be expressed as the sum\nof the Nth powers of unique, natural numbers. For example, if X=13 and N=2.\nWe have to find all combinations of unique squares adding up to 13.\nThe only solution is 2^2+3^2. Constraints: 1<=X<=1000, 2<=N<=10.\n\"\"\"",
      "successors": []
    },
    {
      "id": 2,
      "start_line": 8,
      "end_line": 8,
      "label": "if __name__ == \"__main__\":",
      "successors": [
        {
          "id": 3,
          "start_line": 9,
          "end_line": 10,
          "label": "import doctest\n    doctest.testmod()",
          "successors": []
        }
      ]
    }
  ],
  "functions": [
    {
      "name": "backtrack",
      "type": "CFG",
      "blocks": [
        {
          "id": 1,
          "start_line": 1,
          "end_line": 24,
          "label": "def backtrack(\n    needed_sum: int,\n    power: int,\n    current_number: int,\n    current_sum: int,\n    solutions_count: int,\n) -> tuple[int, int]:\n    \"\"\"\n    >>> backtrack(13, 2, 1, 0, 0)\n    (0, 1)\n    >>> backtrack(10, 2, 1, 0, 0)\n    (0, 1)\n    >>> backtrack(10, 3, 1, 0, 0)\n    (0, 0)\n    >>> backtrack(20, 2, 1, 0, 0)\n    (0, 1)\n    >>> backtrack(15, 10, 1, 0, 0)\n    (0, 0)\n    >>> backtrack(16, 2, 1, 0, 0)\n    (0, 1)\n    >>> backtrack(20, 1, 1, 0, 0)\n    (0, 64)\n    \"\"\"\n    if current_sum == needed_sum:",
          "successors": [
            {
              "id": 2,
              "start_line": 25,
              "end_line": 27,
              "label": "# If the sum of the powers is equal to needed_sum, then we have a solution.\n        solutions_count += 1\n        return current_sum, solutions_count",
              "successors": []
            },
            {
              "id": 3,
              "start_line": 28,
              "end_line": 29,
              "label": "i_to_n = current_number**power\n    if current_sum + i_to_n <= needed_sum:",
              "successors": [
                {
                  "id": 4,
                  "start_line": 30,
                  "end_line": 35,
                  "label": "# If the sum of the powers is less than needed_sum, then continue adding powers.\n        current_sum += i_to_n\n        current_sum, solutions_count = backtrack(\n            needed_sum, power, current_number + 1, current_sum, solutions_count\n        )\n        current_sum -= i_to_n",
                  "successors": []
                }
              ]
            },
            {
              "id": 5,
              "start_line": 36,
              "end_line": 36,
              "label": "if i_to_n < needed_sum:",
              "successors": [
                {
                  "id": 6,
                  "start_line": 37,
                  "end_line": 40,
                  "label": "# If the power of i is less than needed_sum, then try with the next power.\n        current_sum, solutions_count = backtrack(\n            needed_sum, power, current_number + 1, current_sum, solutions_count\n        )",
                  "successors": []
                }
              ]
            }
          ]
        },
        {
          "id": 7,
          "start_line": 41,
          "end_line": 41,
          "label": "return current_sum, solutions_count",
          "successors": []
        }
      ],
      "functions": [],
      "classes": []
    },
    {
      "name": "solve",
      "type": "CFG",
      "blocks": [
        {
          "id": 1,
          "start_line": 1,
          "end_line": 26,
          "label": "def solve(needed_sum: int, power: int) -> int:\n    \"\"\"\n    >>> solve(13, 2)\n    1\n    >>> solve(10, 2)\n    1\n    >>> solve(10, 3)\n    0\n    >>> solve(20, 2)\n    1\n    >>> solve(15, 10)\n    0\n    >>> solve(16, 2)\n    1\n    >>> solve(20, 1)\n    Traceback (most recent call last):\n        ...\n    ValueError: Invalid input\n    needed_sum must be between 1 and 1000, power between 2 and 10.\n    >>> solve(-10, 5)\n    Traceback (most recent call last):\n        ...\n    ValueError: Invalid input\n    needed_sum must be between 1 and 1000, power between 2 and 10.\n    \"\"\"\n    if not (1 <= needed_sum <= 1000 and 2 <= power <= 10):",
          "successors": [
            {
              "id": 2,
              "start_line": 27,
              "end_line": 30,
              "label": "raise ValueError(\n            \"Invalid input\\n\"\n            \"needed_sum must be between 1 and 1000, power between 2 and 10.\"\n        )",
              "successors": []
            },
            {
              "id": 3,
              "start_line": 31,
              "end_line": 31,
              "label": "return backtrack(needed_sum, power, 1, 0, 0)[1]  # Return the solutions_count",
              "successors": []
            }
          ]
        }
      ],
      "functions": [],
      "classes": []
    }
  ],
  "classes": []
}